pkg/plugins/manifest.go

```diff
@@ -83,51 +97,118 @@ func readPluginManifest(body []byte) (*pluginManifest, error) {
 	manifestPath := filepath.Join(plugin.PluginDir, "MANIFEST.txt")
 
 	byteValue, err := ioutil.ReadFile(manifestPath)
 	if err != nil || len(byteValue) < 10 {
 		log.Debug("Plugin is unsigned", "id", plugin.Id)
-		return PluginSignatureUnsigned
+		return &PluginSignatureState{
+			Status: pluginSignatureUnsigned,
+		}, nil
 	}
 
 	manifest, err := readPluginManifest(byteValue)
 	if err != nil {
 		log.Debug("Plugin signature invalid", "id", plugin.Id)
-		return PluginSignatureInvalid
+		return &PluginSignatureState{
+			Status: pluginSignatureInvalid,
+		}, nil
 	}
 
 	// Make sure the versions all match
 	if manifest.Plugin != plugin.Id || manifest.Version != plugin.Info.Version {
-		return PluginSignatureModified
+		return &PluginSignatureState{
+			Status: pluginSignatureModified,
+		}, nil
 	}
 
+	// Validate that private is running within defined root URLs
+	if manifest.SignatureType == privateType {
+		appURL, err := url.Parse(setting.AppUrl)
+		if err != nil {
+			return nil, err
+		}
+
+		foundMatch := false
+		for _, u := range manifest.RootURLs {
+			rootURL, err := url.Parse(u)
+			if err != nil {
+				log.Warn("Could not parse plugin root URL", "plugin", plugin.Id, "rootUrl", rootURL)
+				return nil, err
+			}
+			if rootURL.Scheme == appURL.Scheme &&
+				rootURL.Host == appURL.Host &&
+				rootURL.RequestURI() == appURL.RequestURI() {
+				foundMatch = true
+				break
+			}
+		}
+
+		if !foundMatch {
+			log.Warn("Could not find root URL that matches running application URL", "plugin", plugin.Id, "appUrl", appURL, "rootUrls", manifest.RootURLs)
+			return &PluginSignatureState{
+				Status: pluginSignatureInvalid,
+			}, nil
+		}
+	}
+
+	manifestFiles := make(map[string]bool, len(manifest.Files))
+
 	// Verify the manifest contents
 	log.Debug("Verifying contents of plugin manifest", "plugin", plugin.Id)
 	for p, hash := range manifest.Files {
 		// Open the file
 		fp := filepath.Join(plugin.PluginDir, p)
 		f, err := os.Open(fp)
 		if err != nil {
-			return PluginSignatureModified
+			log.Warn("Plugin file listed in the manifest was not found", "plugin", plugin.Id, "filename", p, "dir", plugin.PluginDir)
+			return &PluginSignatureState{
+				Status: pluginSignatureModified,
+			}, nil
 		}
 		defer f.Close()
 
 		h := sha256.New()
 		if _, err := io.Copy(h, f); err != nil {
 			log.Warn("Couldn't read plugin file", "plugin", plugin.Id, "filename", fp)
-			return PluginSignatureModified
+			return &PluginSignatureState{
+				Status: pluginSignatureModified,
+			}, nil
 		}
 		sum := hex.EncodeToString(h.Sum(nil))
 		if sum != hash {
 			log.Warn("Plugin file's signature has been modified versus manifest", "plugin", plugin.Id, "filename", fp)
-			return PluginSignatureModified
+			return &PluginSignatureState{
+				Status: pluginSignatureModified,
+			}, nil
+		}
+		manifestFiles[p] = true
+	}
+
+	if manifest.isV2() {
+		// Track files that were missing from the manifest
```

Comment:
"Were" makes it sound as if it's in the past:
```suggestion
		// Track files missing from the manifest
```